site/_posts/2020-05-22-building-cli-app-with-device-grant-and-golang.md

```diff
@@ -0,0 +1,303 @@
+	}
+}
+
+// ...
+```
+
+This should look familiar! We start with our LoginCmd's `Run` function again, but we continue that function a bit more to do the following:
+
+1. We call `informUserAndOpenBrowser` with our `deviceResp.UserCode` that we got from the FA server. This prints to the user's CLI the given `userCode` and open's up FA's `/oauth2/device` endpoint for our Application and Tenant.
+1. Next, we call `startPolling` which continually polls the given `openIDConfig.TokenEndpoint` every `retryInterval` seconds (5 seconds is RFC8628's / FA's default) to see if the user has completed the Device login (tied to the given `deviceCode`). Once the user has completed their login, the FA server responds with a proper `200` response including a JWT Access Token which then populates our returned `result` object.
+
+Now we're getting somewhere! The only thing left to do is trade our Access Token for the actual user record and save it to disk. Let's do it!
+
+### Trade our Access Token for info on our User and write it to disk
+
+Our last task is simple: We have a JWT Access Token and our handy `FusionAuthClient` has a method [`RetrieveUserUsingJWT`](https://pkg.go.dev/github.com/FusionAuth/go-client/pkg/fusionauth?tab=doc#FusionAuthClient.RetrieveUserUsingJWT), so we just need to combine the two and save the result. Let's see it in action:
+
+```go
+// LoginCmd provides the subcommand for logging into the FA server using the Device Code grant.
+var LoginCmd = &cobra.Command{
+	Use:   "login [no options!]",
+	Short: "Login to the FA server using the OAuth Device Flow.",
+	Run: func(cmd *cobra.Command, args []string) {
+      // ...
+
+		accessToken, err := startPolling(openIDConfig.TokenEndpoint, deviceResp.DeviceCode, deviceResp.Interval)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		fetchAndSaveUser(accessToken)
+	},
+}
+
+func fetchAndSaveUser(token *fusionauth.AccessToken) {
+	resp, _, err := faClient.RetrieveUserUsingJWT(token.AccessToken)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Save our User object for later usage in fetch
+	Save("/tmp/getgif.json", resp.User)
+
+	cyan := color.New(color.FgCyan)
+	cyan.Printf("You successfully authenticated! You can now use `getgif fetch`!\n")
+}
+```
+
+Great stuff, we send off our JWT to FA, FA's go-client intelligently responds with a [`UserResponse`](https://pkg.go.dev/github.com/FusionAuth/go-client/pkg/fusionauth?tab=doc#UserResponse) object, we save that response's User field (`resp.User`) to `/tmp/getgif.json`, and then we print a success message. Simple, easy, and effective!
+
+## Working!
+
+Now, since this is a post about using FusionAuth and not about the Giphy API, I won't dig into the `fetch` command. I'll leave that to you but suffice to say it's pretty simple. Feel free to [check out the code](https://github.com/FusionAuth/fusionauth-example-go-device-code-grant/blob/master/cmd/fetch.go). Here is the whole flow in action:
+
+<div class="d-flex justify-content-center mb-5 mt-1 youtube">
```

Comment:
The only issue with the video is that it doesn't show the user logging into FusionAuth. Since the person that was recording the video had previously authenticated to FusionAuth, they were redirected to authorize and then immediately redirected to the device success page. If they weren't logged in, they would have been taken to authorize and then logged in as usual. Not a huge issue, but definitely would be nice to show the complete flow.

Reply:
Recorded new video, which actually turned up something that hadn't been documented in the post.